driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -21,116 +21,200 @@
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     */
     DateExpression isDateOr(DateExpression other);
+
+
+    /**
+     * {@code this} value as a {@link ArrayExpression array} if
+     * {@code this} is an array, or the {@code other} array value if
+     * {@code this} is null, or is missing, or is of any other non-array type.
+     *
+     * <p>Warning: this operation does not guarantee type safety. While this
+     * operation is guaranteed to produce an array, the type of the elements of
+     * that array are not guaranteed by the API. The specification of a type by
+     * the user is an unchecked assertion that all elements are of that type,
+     * and that no element is null, is missing, or is of some other type. If the
+     * user cannot make such an assertion, some appropriate super-type should be
+     * chosen, and elements should be individually type-checked.
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     * @param <T> the type of the elements of the resulting array.
+     */
     <T extends Expression> ArrayExpression<T> isArrayOr(ArrayExpression<T> other);
+
+    // TODO-END doc after Map merged, "record" and "schema objects" are decided
     <T extends DocumentExpression> T isDocumentOr(T other);
 
 
     /**
-     * server error if type cannot be converted to string (arrays, objects)
+     * The {@link StringExpression string} value corresponding to this value.
      *
-     * @return
+     * <p>This may "cause an error" if the type cannot be converted to string,
+     * as is the case with {@link ArrayExpression arrays},
+     * {@link DocumentExpression documents}, and {@link MapExpression maps}.
+     * TODO-END what about null/missing?
+     * TODO-END document vs record
+     * TODO-END "cause an error" above
+     *
+     *
+     * @see StringExpression#parseDate()
+     * @see StringExpression#parseInteger()
+     * TODO-END all the others? implement?
+     * @return the resulting value.
      */
     StringExpression asString();
 
     /**
-     * Applies the given function to this argument. Note that "apply" usually
-     * applies functions to arguments; here, the parameters are reversed.
+     * Applies the provided function to {@code this}.
      *
-     * @param f
-     * @return
-     * @param <T>
-     * @param <R>
+     * <p>Equivalent to {@code f.apply(this)}, and allows lambdas and static,
+     * user-defined functions to use the chaining syntax. For example:
+     * {@code myInteger.apply(isEven)} (here, {@code isEven} is a function
+     * taking an {@link IntegerExpression integer} and yielding a
+     * {@link BooleanExpression boolean}).
+     *
+     * @param f the function to apply.
+     * @return the resulting value.
+     * @param <T> the type of {$code this} value.
+     * @param <R> the type of the resulting value.
      */
     <T extends Expression, R extends Expression> R apply(Function<T, R> f);
```

Comment:
The type of `Function<T, R>` is too strict. I keep pointing out problems like this, but they keep coming. This is not productive, and I would hope that the problem stops constantly reappearing in the new code.

Reply:
Resolving because I left the same [comment](https://github.com/mongodb/mongo-java-driver/pull/1055/files#r1064907718) in the PR that introduced the method.